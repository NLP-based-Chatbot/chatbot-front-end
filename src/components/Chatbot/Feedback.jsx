import React, { useState } from 'react'
import { ButtonGroup, Card, Container, IconButton, makeStyles, TextField, Typography } from '@material-ui/core';
import clsx from 'clsx';
import StarIcon from '@material-ui/icons/Star';
import StarBorderIcon from '@material-ui/icons/StarBorder';
import SendIcon from '@material-ui/icons/Send';

const useStyles = makeStyles(theme => ({
  root: {
    height: "100vh",
  },
  card: {
    height: "fit-content",
    padding: theme.spacing(4)
  },
  flex: {
    display: "flex",
    justifyContent: "center",
    alignItems: "center"
  },
  icon_group: {
    marginTop: theme.spacing(2)
  },
  text: {
    width: "100%",
    backgroundColor: theme.palette.secondary.light,
    padding: theme.spacing(1, 2),
  },
  text_area: {
    display: "flex",
    alignItems: "center",
    marginTop: theme.spacing(2)
  },
  text_style: {
    textAlign: "center"
  },
  feedback: {
    width: "100%",
    marginTop: theme.spacing(2)
  }
}))

const Feedback = ({ submit }) => {
  const classes = useStyles()
  const [selected, updateSelected] = useState([0, 0, 0, 0, 0])
  const [feedbackDisplay, setFeedbackDisplay] = useState(false)
  const [feedback, updateFeedback] = useState("")

  const click = (index) => {
    let tempList = [0, 0, 0, 0, 0]
    for (let i = 0; i <= index; i++) {
      tempList[i] = 1
    }
    updateSelected(tempList)

    if (index < 3) {
      setTimeout(() => {
        setFeedbackDisplay(true)
      }, 1000)
    } else {
      setTimeout(() => {
        submit()
      }, 1000)
    }
  }

  return (
    <Container className={clsx(classes.root, classes.flex)}>
      <Card className={clsx(classes.card, classes.flex)} style={{ flexDirection: "column" }}>
        <Typography variant="h6" className={classes.text_style}>Please provide your valueble rating<br />to improve our chatbot</Typography>
        <ButtonGroup className={classes.icon_group}>
          {selected.map((select, index) => {
            return (
              <IconButton key={index} onClick={() => click(index)}>
                {select ? <StarIcon style={{ color: "#EAED15" }} /> : <StarBorderIcon />}
              </IconButton>
            )
          })}
        </ButtonGroup>
        {feedbackDisplay &&
          <div className={classes.feedback}>
            <Typography variant="h6" className={classes.text_style}>Please provide some feedback<br /> on what went wrong</Typography>
            <div className={classes.text_area}>
              <TextField
                className={classes.text}
                multiline={true}
                maxRows={1}
                value={feedback}
                onChange={e => updateFeedback(e.target.value)}
                InputProps={{ disableUnderline: true }}
              />
              <IconButton onClick={() => submit(feedback)}>
                <SendIcon />
              </IconButton>
            </div>
          </div>
        }
      </Card>
    </Container>
  )
}

export default Feedback
