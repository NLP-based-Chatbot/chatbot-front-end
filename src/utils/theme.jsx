import { createTheme } from '@material-ui/core';

const theme = createTheme({
  typography: {
    fontFamily: ["Poppins", "sans-serif"].join(","),
  },
  palette: {
    type: 'light',
    primary: {
      light: "#0292BE",
      main: "#3A637E",
      dark: "#1A2C38",
      contrastText: "#fff",
    },
    secondary: {
      light: "#D3EFF8",
      main: "#BBDB9B",
      dark: "#ABC4A1",
      contrastText: "#000",
    },
  },
});

theme.typography.h6 = {
  [theme.breakpoints.up('md')]: {
    fontSize: "1.5rem",
    fontWeight: 500
  },
  [theme.breakpoints.down('md')]: {
    fontSize: "1.2rem",
    fontWeight: 500
  },
  [theme.breakpoints.down('sm')]: {
    fontSize: "0.8rem",
    fontWeight: 500
  },
  [theme.breakpoints.down('xs')]: {
    fontSize: "0.5rem",
    fontWeight: 500
  },
}

export default theme;