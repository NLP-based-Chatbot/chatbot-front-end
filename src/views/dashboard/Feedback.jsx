import React, { useState, useEffect } from 'react'
import { Box, Button, Card, CardContent, Container, FormControl, FormControlLabel, FormLabel, Grid, makeStyles, Radio, RadioGroup, Typography } from '@material-ui/core';
import api from './../../api/index';
import { useSelector, useDispatch } from 'react-redux';
import { getToken } from './../../store/slices/auth';
import TokenGenerator from './../../helpers/TokenRefresh';
import { refreshToken } from '../../store/slices/auth';
import FeedbackItem from '../../components/FeedbackItem';
import { sessionsReceived, sessionsRequested, sessionsRequestFailed } from '../../store/slices/sessions';
import { getSessions } from '../../store/slices/sessions';
import ChatDisplay from '../../components/Chatbot/ChatDisplay';

const useStyles = makeStyles(theme => ({
  root: {
    background: `linear-gradient(45deg, ${theme.palette.primary.main} 30%, ${theme.palette.primary.dark} 90%)`,
    minHeight: "calc(100vh - 115px)",
    display: "flex",
    alignItems: "center"
  },
  container: {
    height: "80vh"
  },
  breadcrumb: {
    marginBottom: 'auto'
  },
  title: {
    color: theme.palette.primary.contrastText,
  },
  grid_container: {
    height: '100%',
  },
  grid_item1: {
    height: "100%",
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center'
  },
  grid_item2: {
    height: "100%"
  },
  feedback_card: {
    height: "90%",
    width: "100%"
  },
  feedbacks: {
    backgroundColor: '#d2d5d9',
    height: "80%",
    width: "90%",
    borderRadius: "10px",
    overflowY: 'auto'
  },
  heading: {
    marginBottom: theme.spacing(2),
    fontWeight: 600
  },
  subHeading: {
    marginBottom: theme.spacing(2)
  },
  button: {
    marginTop: 'auto',
    marginLeft: 'auto'
  }
}))

const Feedback = () => {
  const classes = useStyles()
  const dispatch = useDispatch()
  const [filter, changeFilter] = useState("all")
  const sessions = useSelector(getSessions)

  const [selected, updateSelected] = useState(null)
  const [selected_feedback, selectFeedback] = useState(null)

  const token = useSelector(getToken)

  useEffect(() => {
    async function loadFeedback() {
      try {
        const access = await TokenGenerator(token)
        if (access) dispatch(refreshToken(access))

        dispatch(sessionsRequested())
        const feedback_data = await api.feedback.GET.sessions(token.access)
        dispatch(sessionsReceived(feedback_data.data))
      } catch (err) {
        console.log(err.message)
        dispatch(sessionsRequestFailed())
      }
    }

    loadFeedback()
  }, [])

  const resolveFeedback = async () => {
    try {
      const result = await api.feedback.PUT.updateFeedback(
        selected_feedback.id,
        selected_feedback.domain,
        selected_feedback.rating,
        selected_feedback.chatsession,
        selected_feedback.feedback,
        selected_feedback.user_id,
        token.access
      )
      window.location.reload(false)
    } catch (err) {
      console.log(err.response)
    }
  }

  useEffect(() => {
    const feedback = sessions.filter(session => session.id === selected)
    selectFeedback(feedback[0])
  }, [selected, updateSelected])

  return (
    <div className={classes.root}>
      <Container className={classes.container} maxWidth='xl'>
        {/* <Breadcrumbs className={classes.breadcrumb}>
          <Link style={{ color: '#d2d5d9' }} href='/admin/dashboard'>
            <Typography variant="h6">Dashboard</Typography>
          </Link>
          <Link style={{ color: 'white' }} href='/admin/dashboard/feedback'>
            <Typography variant="h6">Feedback</Typography>
          </Link>
        </Breadcrumbs> */}
        <Grid container className={classes.grid_container}>
          <Grid item lg={5} className={classes.grid_item1}>
            <Typography variant="h2" className={classes.title}>
              Chat session feedbacks
            </Typography>
          </Grid>
          <Grid item lg={7} className={classes.grid_item1}>
            <Card className={classes.feedback_card}>
              <CardContent style={{ height: '100%' }}>
                <Grid container className={classes.grid_container} justifyContent="space-evenly">
                  <Grid item md={5} className={classes.grid_item1} direction="column">
                    {/* <FormControl component="fieldset">
                      <FormLabel component="legend">Filter Feedback</FormLabel>
                      <RadioGroup name="filter" value={filter} onChange={e => changeFilter(e.target.value)}>
                        <FormControlLabel control={<Radio />} label="All" value="all" />
                        <FormControlLabel control={<Radio />} label="Negative Feedback" value="negative" />
                        <FormControlLabel control={<Radio />} label="Positive Feedback" value="positive" />
                        <FormControlLabel control={<Radio />} label="Resolved" value="resolved" />
                        <FormControlLabel control={<Radio />} label="Unresolved" value="unresolved" />
                      </RadioGroup>
                    </FormControl> */}

                    <Box className={classes.feedbacks}>
                      {sessions && sessions.map((session, index) => <FeedbackItem key={index} index={index + 1} selectFeedback={(id) => updateSelected(id)} {...session} />)}
                    </Box>
                  </Grid>
                  <div style={{ border: '1px solid #d2d5d9', height: '100%' }} />
                  <Grid item md={6} direction='column' className={classes.grid_item2}>
                    <Typography variant="h5" className={classes.heading}>Feedback Details</Typography>
                    <Grid container>
                      <Grid item md={3}>
                        <Typography variant="h6" style={{ fontWeight: 600 }}>Domain</Typography>
                        {selected && selected_feedback && <Typography variant="h6">{selected_feedback.domain.toUpperCase()}</Typography>}
                      </Grid>
                      <Grid item md={6}>
                        <Typography variant="h6" style={{ fontWeight: 600 }}>Resolve Status</Typography>
                        {selected && selected_feedback && <Typography variant="h6">{selected_feedback.resolved ? 'Resolved' : 'Unresolved'}</Typography>}
                      </Grid>
                      <Grid item md={3}>
                        <Typography variant="h6" style={{ fontWeight: 600 }}>Rating</Typography>
                        {selected && selected_feedback && <Typography variant="h6">{selected_feedback.rating}</Typography>}
                      </Grid>
                    </Grid>
                    <ChatDisplay />
                    {selected && selected_feedback && !selected_feedback.resolved &&
                      <Button
                        variant='contained'
                        color='secondary'
                        onClick={() => resolveFeedback()}
                        className={classes.button}
                      >
                        Mark As Resovled
                      </Button>}
                  </Grid>
                </Grid>
              </CardContent>
            </Card>
          </Grid>
        </Grid>
      </Container>
    </div>
  )
}

export default Feedback
