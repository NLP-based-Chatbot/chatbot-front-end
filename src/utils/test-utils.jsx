import React from "react";
import { render } from "@testing-library/react";
import { ThemeProvider } from "@material-ui/core";
import theme from "./theme";
import { ParallaxProvider } from "react-scroll-parallax";
import { BrowserRouter } from "react-router-dom";
import "@babel/polyfill";
import { configureStore } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
import AuthReducer from '../store/slices/auth';

const customRender = (
  ui,
  {
    preloadedState,
    store = configureStore({ reducer: { auth: AuthReducer }, preloadedState }),
    ...renderOptions
  } = {}
) => {
  const AllTheProviders = ({ children }) => {
    return (
      <Provider store={store}>
        <BrowserRouter>
          <ParallaxProvider>
            <ThemeProvider theme={theme}>{children}</ThemeProvider>;
          </ParallaxProvider>
        </BrowserRouter>
      </Provider>
    );
  };
  return render(ui, { wrapper: AllTheProviders, ...renderOptions });
};

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
