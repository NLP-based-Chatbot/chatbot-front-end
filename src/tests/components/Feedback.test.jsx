/**
 * @jest-environment jsdom
 */

import { render } from "../../utils/test-utils";
import { screen } from "@testing-library/react";
import React from "react";
import userEvent from "@testing-library/user-event";
import "@testing-library/jest-dom/extend-expect";
import Feedback from "../../components/Chatbot/Feedback";
import renderer from "react-test-renderer";

test("initial render", async () => {
  render(<Feedback />);
  screen.debug();
});

test("prompt for feedback when rating is below 3", async () => {
  const handleSubmit = jest.fn();
  render(<Feedback submit={handleSubmit} />);
  userEvent.click(screen.getByTestId("star-1"));
  
  expect(await screen.findByTestId("feedback-form")).toBeInTheDocument()
  //screen.debug()
});

test("submit when rating is above 3", async () => {
  const handleSubmit = jest.fn();
  render(<Feedback submit={(index, feedback) => handleSubmit(index, feedback)} />);
  userEvent.click(screen.getByTestId("star-5"));
  
  await new Promise((r) => setTimeout(r, 1000));
  expect(handleSubmit).toHaveBeenCalledWith(5,'none')
  //screen.debug()
});

test("submit feedback when rating is below 3", async () => {
  const handleSubmit = jest.fn();
  render(<Feedback submit={handleSubmit} />);
  userEvent.click(screen.getByTestId("star-1"));
  
  expect(await screen.findByTestId("feedback-form")).toBeInTheDocument()
  userEvent.type(screen.getByTestId('feedback-text'), 'test feedback')
  userEvent.click(screen.getByTestId('submit-btn'))
  expect(handleSubmit).toHaveBeenCalledWith(1,'test feedback')
  //screen.debug()
});

test("snapshot test", () => {
  const tree = renderer.create(<Feedback />).toJSON();
  expect(tree).toMatchSnapshot();
  //screen.debug();
});