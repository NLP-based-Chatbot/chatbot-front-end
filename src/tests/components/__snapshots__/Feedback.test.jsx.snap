// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test 1`] = `
<div
  className="MuiContainer-root makeStyles-root-33 makeStyles-flex-35 MuiContainer-maxWidthLg"
>
  <div
    className="MuiPaper-root MuiCard-root makeStyles-card-34 makeStyles-flex-35 MuiPaper-elevation1 MuiPaper-rounded"
    style={
      Object {
        "flexDirection": "column",
      }
    }
  >
    <h6
      className="MuiTypography-root makeStyles-text_style-39 MuiTypography-h6"
    >
      Please provide your valueble rating
      <br />
      to improve our chatbot
    </h6>
    <div
      className="MuiButtonGroup-root makeStyles-icon_group-36"
      role="group"
    >
      <button
        className="MuiButtonBase-root MuiIconButton-root MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlined"
        data-testid="star-1"
        disableElevation={false}
        disabled={false}
        fullWidth={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
        variant="outlined"
      >
        <span
          className="MuiIconButton-label"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M22 9.24l-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03L22 9.24zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28L12 15.4z"
            />
          </svg>
        </span>
      </button>
      <button
        className="MuiButtonBase-root MuiIconButton-root MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlined"
        data-testid="star-2"
        disableElevation={false}
        disabled={false}
        fullWidth={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
        variant="outlined"
      >
        <span
          className="MuiIconButton-label"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M22 9.24l-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03L22 9.24zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28L12 15.4z"
            />
          </svg>
        </span>
      </button>
      <button
        className="MuiButtonBase-root MuiIconButton-root MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlined"
        data-testid="star-3"
        disableElevation={false}
        disabled={false}
        fullWidth={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
        variant="outlined"
      >
        <span
          className="MuiIconButton-label"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M22 9.24l-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03L22 9.24zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28L12 15.4z"
            />
          </svg>
        </span>
      </button>
      <button
        className="MuiButtonBase-root MuiIconButton-root MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlined"
        data-testid="star-4"
        disableElevation={false}
        disabled={false}
        fullWidth={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
        variant="outlined"
      >
        <span
          className="MuiIconButton-label"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M22 9.24l-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03L22 9.24zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28L12 15.4z"
            />
          </svg>
        </span>
      </button>
      <button
        className="MuiButtonBase-root MuiIconButton-root MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlined"
        data-testid="star-5"
        disableElevation={false}
        disabled={false}
        fullWidth={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
        variant="outlined"
      >
        <span
          className="MuiIconButton-label"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M22 9.24l-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03L22 9.24zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28L12 15.4z"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>
</div>
`;
